File: solutions/_290_word_pattern/solution.java
package _290_word_pattern;

import java.util.*;

class Solution {

  public boolean wordPattern(String pattern, String s) {
    String[] words = s.split(" ");
    if (words.length != pattern.length()) return false;

    Map<Character, String> map = new HashMap<>();
    Set<String> set = new HashSet<>();

    for (int i = 0; i < pattern.length(); i++) {
      char c = pattern.charAt(i);
      String word = words[i];

      if (map.containsKey(c)) {
        if (!map.get(c).equals(word)) return false;
      } else {
        if (set.contains(word)) return false;
        map.put(c, word);
        set.add(word);
      }
    }

    return true;
  }
}
/*
https://leetcode.com/problems/word-pattern/
*/

File: solutions/_290_word_pattern/solution.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split()
        if len(words) != len(pattern):
            return False
        pattern_to_word = {}
        word_to_pattern = {}
        for i, word in enumerate(words):
            if pattern[i] in pattern_to_word:
                if pattern_to_word[pattern[i]] != word:
                    return False
            else:
                pattern_to_word[pattern[i]] = word
            if word in word_to_pattern:
                if word_to_pattern[word] != pattern[i]:
                    return False
            else:
                word_to_pattern[word] = pattern[i]
        return True


# https://leetcode.com/problems/word-pattern/
